keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,643 @@
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+<!--
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
+  - Metric name:
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+<!--
+Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+implementation difficulties, etc.).
+-->
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### Does this feature depend on any specific services running in the cluster?
+
+<!--
+Think about both cluster-level services (e.g. metrics-server) as well
+as node-level agents (e.g. specific version of CRI). Focus on external or
+optional services that are needed. For example, if this feature depends on
+a cloud provider API, or upon an external software-defined storage or network
+control plane.
+
+For each of these, fill in the following—thinking about running existing user workloads
+and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]
+    - Usage description:
+      - Impact of its outage on the feature:
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+### Scalability
+
+<!--
+For alpha, this section is encouraged: reviewers should consider these questions
+and attempt to answer them.
+
+For beta, this section is required: reviewers must answer these questions.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### Will enabling / using this feature result in any new API calls?
+
+Not directly, but users or controllers may call more deletion requests to retry.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+Describe them, providing:
+  - API type(s): `ObjectMeta`
+  - Estimated increase in size: a slice of `Liens` of size (TBD) B
```

Comment:
I think we should ensure that the lienID is no longer than N bytes (64? 256?).

The separate question is if we should force upper bound for the size of the slice...

Reply:
> I think we should ensure that the lienID is no longer than N bytes (64? 256?).

Thank you for suggestion.

Length of each key will be about 32 to 64 bytes.
And the number of keys may be around 4 to 8 at most.

Therefore, 256B will be nice as an estimated size (For most case, it will be 0 to 64B). Updated the KEP with this value.

> The separate question is if we should force upper bound for the size of the slice...

As for current finalizers, the only [validation](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L350) seems checking if all the keys are namespaced. So, it seems to have been working well without upper bound until now (I don't mean to say that upper bound isn't necessary just because of it).

If we would like to introduce per key upper bound length, we should add it before beta because it will break compatibility.
As for the upper bound for length of the array or max size of array, I'm not sure that I can set a hard limit in a safe manner (It will introduce inconsistency, because which key can't be added depends on the order of addition).